en/middleware/micrortps_throughput_test.md

```diff
@@ -0,0 +1,105 @@
+
+```text
+uint8[256] data
+```
+
+This can be done with the command line below:
+
+```sh
+cd /path/to/PX4/Firmware/msg
+echo uint8[256] data > throughput_256.msg
+```
+
+Register the new message adding it to the list of messages in the file: **/Firmware/msg/CMakeLists.txt**:
+
+```cmake
+...
+wind_estimate.msg
+throughput_256.msg
+)
+...
+```
+
+Give the message a topic id by adding a line in the **/Firmware/Tools/message_id.py** script:
+
+```python
+...
+    'wind_estimate': 94,
+    'throughput_256': 95,
+}
+...
+```
+
+## Disable automatic bridge code generation
+
+Disable automatic generation of bridge code (as part of the PX4 build process) by setting the variable `GENERATE_RTPS_BRIDGE` to `off` in the *.cmake* file for the target platform (*cmake/configs/*):
+
+```sh
+set(GENERATE_RTPS_BRIDGE off)
+```
+
+
+## Generate the bridge code
+
+Manually generate bridge code using *generate_microRTPS_bridge.py* (the code will send and receive "just" our `throughput_256` uORB topic):
+
+```sh
+cd /path/to/PX4/Firmware
+python Tools/generate_microRTPS_bridge.py --send msg/throughput_256.msg --receive msg/throughput_256.msg
+```
+
+The *Client* source code is generated in **src/modules/micrortps_bridge/micrortps_client/** and the *Agent* in **src/modules/micrortps_bridge/micrortps_agent/**.
+
+### Modify the client code
+
+Next we modify the *Client* to send a *throughput_256* message on every loop. This is required because the topic is not actually being published by PX4, and because we want to ensure that it is sent at the greatest possible rate.
+
+Open the file **src/modules/micrortps_bridge/micrortps_client/microRTPS_client.cpp**. Update the `while` loop in the `send()` function to look like this:
+
+```cpp
+...
+while (!_should_exit_task)
+{
+    //bool updated;
+    //orb_check(fds[0], &updated);
+    //if (updated)
+    {
+        // obtained data for the file descriptor
+        struct throughput_256_s data = {};
+        // copy raw data into local buffer
+        //orb_copy(ORB_ID(throughput_256), fds[0], &data);
+        data.data[0] = loop%256;
+        serialize_throughput_256(&data, data_buffer, &length, &microCDRWriter);
+        if (0 < (read = transport_node->write((char)95, data_buffer, length)))
+        {
+            total_sent += read;
+            ++sent;
+        }
+    }
+     usleep(_options.sleep_ms*1000);
+    ++loop;
+}
+...
+```
+
+Compile and launch both the [client](../middleware/micrortps.md#client-px4-firmware) and the [agent](../middleware/micrortps.md#agent-off-board-fastrtps-interface).
+
+## Result
+
+The throughput can be observed in the client shell window.
+
+For PX4 running on Pixracer connected to an ordinary PC running Ubuntu 16.04, the results are:
+
+```sh
+SENT:     13255 messages in 13255 LOOPS, 3512575 bytes in 30.994 seconds - 113.33KB/s
+RECEIVED: 13251 messages in 10000 LOOPS, 3511515 bytes in 30.994 seconds - 113.30KB/s
```

Comment:
No. In this case the agent have a RTPS publisher (receiving from the link) and a RTPS subscriber (sending through the link) over the same topic so the info published is received for the subscriber.

This is a test for the bridge that includes a completelly new message, uORB doesn't know it so it's ignored. For a general case, all the info published over the same topic is received for all the subscribers of this topic. The publisher and the subscriber are different entities although can be in the same application.